src/main/java/com/spotify/confidence/EventSenderEngineImpl.java

```diff
@@ -1,162 +1,173 @@
+      new ConcurrentLinkedQueue<>();
+  private final Set<CompletableFuture<?>> pendingBatches = ConcurrentHashMap.newKeySet();
+  private final Thread pollingThread = new Thread(this::pollLoop);
+  private final long maxMemoryConsumption;
+  private volatile boolean intakeClosed = false;
+  private final AtomicLong estimatedMemoryConsumption = new AtomicLong(0);
 
+  @VisibleForTesting
   EventSenderEngineImpl(
-      int maxBatchSize, EventUploader eventUploader, Clock clock, int flushTimeoutMilliseconds) {
-    this.maxBatchSize = maxBatchSize;
+      int maxBatchSize,
+      EventUploader eventUploader,
+      Clock clock,
+      Duration maxFlushInterval,
+      long maxMemoryConsumption) {
     this.eventUploader = eventUploader;
     this.clock = clock;
-    this.flushTimeoutMilliseconds = flushTimeoutMilliseconds;
-    writeThread.submit(new WritePoller());
-    uploadThread.submit(new UploadPoller());
+    this.maxBatchSize = maxBatchSize;
+    this.maxFlushInterval = maxFlushInterval;
+    this.maxMemoryConsumption = maxMemoryConsumption;
+    uploadExecutor =
+        Failsafe.with(
+            RetryPolicy.<Boolean>builder()
+                .handleResult(false)
+                .withBackoff(1, 10, ChronoUnit.SECONDS)
+                .withJitter(0.1)
+                .withMaxAttempts(-1)
+                .withMaxDuration(Duration.ofMinutes(30))
+                .build());
+    pollingThread.start();
   }
 
-  class WritePoller implements Runnable {
-    @Override
-    public void run() {
-      while (true) {
-        final Event event = writeQueue.poll();
-        if (event == null && isStopped) {
-          shutdownQueue.add(SHUTDOWN_WRITE_COMPLETED);
-          break;
-        } else if (event == null) {
-          Thread.yield();
-          continue;
-        }
-        final var numberOfPendingEvents = eventStorage.write(event);
-        if (numberOfPendingEvents >= maxBatchSize) {
-          flush();
-        } else {
-          scheduleFlush();
-        }
-      }
-    }
+  EventSenderEngineImpl(String clientSecret, ManagedChannel channel, Clock clock) {
+    this(
+        DEFAULT_BATCH_SIZE,
+        new GrpcEventUploader(clientSecret, clock, channel),
+        clock,
+        DEFAULT_MAX_FLUSH_INTERVAL,
+        DEFAULT_MAX_MEMORY_CONSUMPTION);
   }
 
-  private void scheduleFlush() {
-    // Cancel the existing scheduled task if it exists
-    if (pendingFlush != null && !pendingFlush.isDone()) {
-      pendingFlush.cancel(false);
-    }
-    if (flushTimeoutMilliseconds > 0) {
-      pendingFlush =
-          executorService.schedule(this::flush, flushTimeoutMilliseconds, TimeUnit.MILLISECONDS);
-    }
+  static Event.Builder event(
+      String name, ConfidenceValue.Struct context, Optional<ConfidenceValue.Struct> message) {
+    return com.spotify.confidence.events.v1.Event.newBuilder()
+        .setEventDefinition(EVENT_NAME_PREFIX + name)
+        .setPayload(
```

Comment:
Should we reverse the order of these? Shouldn't message overwrite context?

Reply:
Actually no, with this approach, if the user adds a `context` key in the message, we will overwrite that with the actual context. We want that, not the other way around.